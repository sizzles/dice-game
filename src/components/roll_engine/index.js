import MersenneTwister from 'mersenne-twister';

//Using a pseudo random number generator allows this core bit of logic to be testable
export class RollEngine {
    constructor(seed) {

      if (typeof seed === undefined || seed === null) {
        this.generator =  new MersenneTwister();
      }

      else {
        this.generator = new MersenneTwister(seed);
      }
    }

    GetRoll() {
        let rand = this.generator.random();
        return Math.ceil(rand * 6);
    }
}