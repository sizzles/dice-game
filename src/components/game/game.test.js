import React from 'react';
import renderer from 'react-test-renderer';
import Game from './index'
import {render, fireEvent, screen} from '@testing-library/react'
import { RollEngine } from '../roll_engine';

const test_seed = 77;

//First few rounds of seed 77 as follows:
//player win
//player win
//player win
//computer win
//computer win
//computer win
//computer win
//computer win
//computer win
//computer win
//draw
//player win

it('renders a new game correctly', () => {

    let rollEngine = new RollEngine(test_seed);

    const tree = renderer
        .create(<Game rollEngine={rollEngine}></Game>)
        .toJSON();
    expect(tree).toMatchSnapshot();
})

async function rollAndCheckResult(text) {
    fireEvent.click(screen.getByTestId("rollButton"))
    const items = await screen.findAllByText(text)
    expect(items).toHaveLength(1)
}

it('plays rounds correctly', async () => {

    let rollEngine = new RollEngine(test_seed);
    render(<Game rollEngine={rollEngine}></Game>)
    await rollAndCheckResult("Player Wins!!!")
    await rollAndCheckResult("Player Wins!!!")
    await rollAndCheckResult("Player Wins!!!")
    await rollAndCheckResult("Computer Wins!!!")
    await rollAndCheckResult("Computer Wins!!!")
    await rollAndCheckResult("Computer Wins!!!")
    await rollAndCheckResult("Computer Wins!!!")
    await rollAndCheckResult("Computer Wins!!!")
    await rollAndCheckResult("Computer Wins!!!")
    await rollAndCheckResult("Computer Wins!!!")
    await rollAndCheckResult("Draw!!!")
})

it('resets the game correctly', async () => { 

    let rollEngine = new RollEngine(test_seed);
    render(<Game  rollEngine={rollEngine}></Game>)

    //Play 1 round then reset
    await rollAndCheckResult("Player Wins!!!")
    fireEvent.click(screen.getByTestId("resetButton"))
    const items = await screen.findAllByText("Let's Play!")
    expect(items).toHaveLength(1)
})