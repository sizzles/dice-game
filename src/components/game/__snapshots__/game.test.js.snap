// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a new game correctly 1`] = `
<div
  className="Game"
>
  <div
    className="Player"
  >
    You
    <div
      className="Dice"
    >
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-dice fa-w-20 "
        data-icon="dice"
        data-prefix="fas"
        focusable="false"
        role="img"
        style={Object {}}
        viewBox="0 0 640 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M592 192H473.26c12.69 29.59 7.12 65.2-17 89.32L320 417.58V464c0 26.51 21.49 48 48 48h224c26.51 0 48-21.49 48-48V240c0-26.51-21.49-48-48-48zM480 376c-13.25 0-24-10.75-24-24 0-13.26 10.75-24 24-24s24 10.74 24 24c0 13.25-10.75 24-24 24zm-46.37-186.7L258.7 14.37c-19.16-19.16-50.23-19.16-69.39 0L14.37 189.3c-19.16 19.16-19.16 50.23 0 69.39L189.3 433.63c19.16 19.16 50.23 19.16 69.39 0L433.63 258.7c19.16-19.17 19.16-50.24 0-69.4zM96 248c-13.25 0-24-10.75-24-24 0-13.26 10.75-24 24-24s24 10.74 24 24c0 13.25-10.75 24-24 24zm128 128c-13.25 0-24-10.75-24-24 0-13.26 10.75-24 24-24s24 10.74 24 24c0 13.25-10.75 24-24 24zm0-128c-13.25 0-24-10.75-24-24 0-13.26 10.75-24 24-24s24 10.74 24 24c0 13.25-10.75 24-24 24zm0-128c-13.25 0-24-10.75-24-24 0-13.26 10.75-24 24-24s24 10.74 24 24c0 13.25-10.75 24-24 24zm128 128c-13.25 0-24-10.75-24-24 0-13.26 10.75-24 24-24s24 10.74 24 24c0 13.25-10.75 24-24 24z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
    </div>
    <div>
      Wins: 0
    </div>
  </div>
  <div>
    <div
      className="RollOutcome"
    >
      Let's Play!
    </div>
    <div>
      <button
        className="GameButton"
        data-testid="rollButton"
        onClick={[Function]}
      >
        Roll Dice
      </button>
    </div>
    <div>
      <button
        className="GameButton"
        data-testid="resetButton"
        onClick={[Function]}
      >
        Reset
      </button>
    </div>
  </div>
  <div
    className="Player"
  >
    Computer
    <div
      className="Dice"
    >
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-dice fa-w-20 "
        data-icon="dice"
        data-prefix="fas"
        focusable="false"
        role="img"
        style={Object {}}
        viewBox="0 0 640 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M592 192H473.26c12.69 29.59 7.12 65.2-17 89.32L320 417.58V464c0 26.51 21.49 48 48 48h224c26.51 0 48-21.49 48-48V240c0-26.51-21.49-48-48-48zM480 376c-13.25 0-24-10.75-24-24 0-13.26 10.75-24 24-24s24 10.74 24 24c0 13.25-10.75 24-24 24zm-46.37-186.7L258.7 14.37c-19.16-19.16-50.23-19.16-69.39 0L14.37 189.3c-19.16 19.16-19.16 50.23 0 69.39L189.3 433.63c19.16 19.16 50.23 19.16 69.39 0L433.63 258.7c19.16-19.17 19.16-50.24 0-69.4zM96 248c-13.25 0-24-10.75-24-24 0-13.26 10.75-24 24-24s24 10.74 24 24c0 13.25-10.75 24-24 24zm128 128c-13.25 0-24-10.75-24-24 0-13.26 10.75-24 24-24s24 10.74 24 24c0 13.25-10.75 24-24 24zm0-128c-13.25 0-24-10.75-24-24 0-13.26 10.75-24 24-24s24 10.74 24 24c0 13.25-10.75 24-24 24zm0-128c-13.25 0-24-10.75-24-24 0-13.26 10.75-24 24-24s24 10.74 24 24c0 13.25-10.75 24-24 24zm128 128c-13.25 0-24-10.75-24-24 0-13.26 10.75-24 24-24s24 10.74 24 24c0 13.25-10.75 24-24 24z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
    </div>
    <div>
      Wins: 0
    </div>
  </div>
</div>
`;
